core/src/com/unciv/logic/multiplayer/chat/ChatStore.kt

```diff
@@ -0,0 +1,96 @@
+package com.unciv.logic.multiplayer.chat
+
+import com.badlogic.gdx.Gdx
+import com.unciv.ui.screens.worldscreen.chat.ChatPopup
+import java.util.LinkedList
+
+data class Chat(
+    val gameId: String,
+) {
+    // <civName, message> pairs
+    private val messages: MutableList<Pair<String, String>> = mutableListOf(INITIAL_MESSAGE)
+
+    val length: Int get() = messages.size
+
+    /**
+     * Only requests a message to be sent.
+     * Does not guarantee delivery.
+     * Failures are mosly ignored.
+     *
+     * The server will relay it back if a delivery was acknowledged and that is when we should display it.
+     */
+    fun requestMessageSend(civName: String, message: String) {
+        Gdx.app.postRunnable {
+            ChatWebSocket.requestMessageSend(Message.Chat(gameId, civName, message))
+        }
+    }
+
+    /**
+     * Although public, this should only be called when a ChatMessageReceivedEvent is received once.
+     */
+    fun addMessage(civName: String, message: String) {
+        messages.add(Pair(civName, message))
+    }
+
+    fun forEachMessage(action: (String, String) -> Unit) {
+        for ((civName, message) in messages) {
+            action(civName, message)
+        }
+    }
+
+    companion object {
+        val INITIAL_MESSAGE = "System" to "Welcome to Chat!"
+    }
+}
+
+object ChatStore {
+    /** The reference to the [ChatPopup] instance which trying to take a peek at our messages currently.
+     * What audacity this popup has!!
+     */
+    var chatPopup: ChatPopup? = null
+
+    private var gameIdToChat = mutableMapOf<String, Chat>()
+
+    /** When no [ChatPopup] is open to receive these oddities, we keep them here.
+     * Certainly better than not knowing why the socket closed.
+     */
+    private var globalMessages = LinkedList<Pair<String, String>>()
```

Comment:
Why a linked list and not an array list? They're more efficient for the uses I see here, we don't remove from the middle nor traverse partway

Reply:
??
Insert is not slow, it's faster on average that linkedlist because the memory allocation is done in batches

Reply:
And we also don't need poll() or removal of a single one. What's happening here is iterating on the entire list, so we can iterate and clear.
What you actually have here is a usecase for a *queue*, not a linked list.